### PR TITLE
Feat: Logging alternative to #547

### PR DESCRIPTION
**Pull request by Kigstn**
## What type of pull request is this?

- [x] Non-breaking code change
- [ ] Breaking code change
- [ ] Documentation change/addition
- [ ] Tests change

## Description

This is an alternative solution to #547, attempting to solve the logging problem.

This PR allows users to pass their own logger at client initialisation. The passed logger is saved in both the client and naff.constants.

This has several benefits;
1) users don't have to do nasty stuff at the very start of their code to overwrite the naff logger before naff does anything with it
2) custom loggers, like `rich.logging` or file handlers can be easily used
3) users have more direct and easier access to the logger via both `client.logger` and `naff.consts.logger`, instead of having to `logging.getLogger(logger_name)` every time they want to use it
4) naff devs have an easier time as well, since they can import the logger from anywhere as well

## Changes

- a logger obj is now defined in `naff.consts` which is overwritten by the logger the user passes to `Client`

## Checklist

- [x] I've formatted my code with [Black](https://black.readthedocs.io/en/stable/)
- [x] I've ensured my code works on `Python 3.10.x`
- [x] I've tested my code

### CODE REVIEW

File: naff/client/client.py [hunk 1]
```diff
@@ -193,8 +195,12 @@ def __init__(
         **kwargs,
     ) -> None:
 
-        # Configuration
+        # Set Up logger and overwrite the constant
+        self.logger = logger
+        """The logger NAFF should use"""
+        constants.logger = logger
```
Comment (silasary) [inline naff/client/client.py:208]: How does this interact with multiple clients in the same process?
Can I define a separate logger for each, and will it actually work as expected?

(I assume not, but I feel the need to raise the question)

Comment (Kigstn) [reply naff/client/client.py:208]: Only one logger is supported, it won't work. Theoretically it should be possible by passing the client obj around to literally every function, but that would be a lot of work to make that change for a very niche feature

Comment (LordOfPolls) [reply naff/client/client.py:208]: Pending resolution of this comment before i can merge

Comment (silasary) [reply naff/client/client.py:208]: I'm fine with the answer being "this isn't supported". But a simple mention of that fact in the docs would be appreciated

**Comment by LordOfPolls** [comment]
On-Hold pending discussion and conflict resolution

File: naff/client/client.py (hunk 1 above)
Comment (Kigstn) [reply naff/client/client.py:208]: Good point, I will put that in the docstring

**Comment by Kigstn** [comment]
Ready for review

**Review by LordOfPolls — APPROVED** [review]
It hurts me calling it `logger`

**Comment by Kigstn** [comment]
> It hurts me calling it `logger`

It hurts me calling it `log`

**Pull request merged**